package/base-files/files/lib/functions/ipv4.sh

```diff
@@ -0,0 +1,166 @@
+uint_max=4294967295
+
+# Common error handler
+__die() {
```

Comment:
I've never really been happy with the `exit` inside shell library code.
In a work project, I would generally not allow it. Symbols from a shell library should also have a common prefix, etc. With free software mostly being volunteer work, I try not to nag too much.

But I'm bringing it up again because it hinders error handling.
It's not possible to catch an error unless the function is called in a subshell (which you don't suspect you'd have to do).

With ipcalc.sh in-between and most functions at least going through `$()` (which isn't run in a new process but handles the `exit`) it isn't hugely dramatic **now** but if anyone either merges the range the calculation into `ipv4.sh` (with callers no longer going through ipcalc.sh) or decides to use any other library functions directly, their processes just die.

I started "fumbling with the exceptionally stable ipcalc.sh" ('stable' for not having been maintained in x years) because a mistake in the configuration of an (otherwise unrelated) interface (like `lan`) can bring down local name resolution on the router itself (and, thus, an important part of the internet connection).
In fact, a configuration or runtime error on a separate interface (e.g. `guest`) could bring down `br-lan` and the name resolution.

Either it has to be absolutely clear that `check_uint32`, `ip2int`, `netmask2prefix`, and `netmask2prefix` can cause the process to exit (which is why I proposed the `assert_` prefix) or these functions have to fail with a return value (my preferred option because otherwise, there'd have to non-`assert_` alternatives for them sooner or later).
`ipv4.sh` looks very juicy and I suspect other packages want to use it as well - we can't assume here that they're all single-purpose and that calling exit has no side effects.

Reply:
> It's not possible to catch an error unless the function is called in a subshell (which you don't suspect you'd have to do).

Actually, if I'm calling `func args` and ignoring the return value, then that would run in the shell.

But if I'm calling `$(func args)` and capturing the output, then that runs in a sub shell.  All of the functions in `ipv4.sh` *except* `is_contiguous()` are expected to run in a sub shell.  That's why `is_contiguous()` returns true/false but never exits.

> With `ipcalc.sh` in-between and most functions at least going through `$()` (which isn't run in a new process [...])

Again, all but one of the functions will get called in a sub shell.

>  [...] or decides to use any other library functions directly, their processes just die.

Again, that depends on whether they're using `eval` or `$(...)` to call those functions.

I could have functions that return their results by polluting the callers environment (and hence never have to run in a sub shell, because they generate no output) but you didn't seem to like that idea either.

>` ipv4.sh` looks very juicy and I suspect other packages want to use it as well - we can't assume here that they're all single-purpose and that calling exit has no side effects.

Indeed, I wrote it because I'll need to use it to do the rewrite of `dhcpd.init` in `isc-dhcp` to output Kea configs instead.

Reply:
> Actually, if I'm calling `func args` [..] then that would run in the shell.

I'm well aware.
It's the

> and ignoring the return value

that I'm worried about.

Here's an example that doesn't ignore the return value:

```shell
f() { exit 1; }
f || echo unreachable
```

Here's a (imo) reasonable way to use `check_uint32`:

```shell
check_uint32 "-1" || {
  # imposible to reach
}
```

> I could have functions that return their results by polluting the callers environment (and hence never have to run in a sub shell, because they generate no output) but you didn't seem to like that idea either.

I'm not sure what you mean. Yes, I don't like library functions terminating my programs unexpectedly AND I don't like them unexpectedly overwriting my variables or functions too.
I'm convinced that is neither unreasonable to want nor impossible to achieve.

> All of the functions in ipv4.sh except is_contiguous() are expected to run in a sub shell.

That fact would at least have to be  be documented somewhere but notes like that are easily overlooked.
If that is true, you can use `f() ( .. )` (rather than `f() { .. }`) to define those functions and the problem goes away.
People working on ipv4.sh in the future should be made aware as well, which I hope a comment on `__die` suffices for.

```shell
f() ( exit 1; )
f || echo reachable
```

Reply:
> > Actually, if I'm calling `func args` [..] then that would run in the shell.
> 
> I'm well aware. It's the
> 
> > and ignoring the return value
> 
> that I'm worried about.
> 
> Here's an example that doesn't ignore the return value:
> 
> ```shell
> f() { exit 1; }
> f || echo unreachable
> ```

This is a red herring.

The only functions that `exit 1` are the ones that generate results not by changing external state, but merely by outputting their calculation on `stdout`.  They will *always* be called via `$(...)` and hence as a sub shell.

The example you cite above just isn't how they will be invoked.

> Here's a (imo) reasonable way to use `check_uint32`:
> 
> ```shell
> check_uint32 "-1" || {
>   # imposible to reach
> }
> ```

Per your suggestion, I changed this to (1) `assert_uint32` and (2) it does a `return 1` not an `exit 1`.  That change was already present when you made this comment.

>  > I could have functions that return their results by polluting the callers environment (and hence never have to run in a sub shell, because they generate no output) but you didn't seem to like that idea either.
> 
> I'm not sure what you mean. Yes, I don't like library functions terminating my programs unexpectedly AND I don't like them unexpectedly overwriting my variables or functions too. I'm convinced that is neither unreasonable to want nor impossible to achieve.

The problem is that you have a scarcity of means here.  *If* we're putting the result of a calculation onto `stdout`, then by necessity we're running the function in a sub shell, *and* we need a means to generate an error that doesn't overload `stdout`.  This is shell, a 40 year old language.  We're not going to find a previously overlooked methodology for handling this usage case.

The mechanism to communicate errors in sub shells that doesn't compromise `stdout` *is* the `exit` status.

> > All of the functions in ipv4.sh except is_contiguous() are expected to run in a sub shell.
> 
> That fact would at least have to be be documented somewhere but notes like that are easily overlooked. If that is true, you can use `f() ( .. )` (rather than `f() { .. }`) to define those functions and the problem goes away. People working on ipv4.sh in the future should be made aware as well, which I hope a comment on `__die` suffices for.
> 
> ```shell
> f() ( exit 1; )
> f || echo reachable
> ```

Reply:
> The mechanism to communicate errors in sub shells that doesn't compromise stdout is the exit status.

'in sub shells". If there is no subshell, the argument is moot. I'm trying to tell you that it's hard to control for that reliably for users now and future code with the current code.

> Per your suggestion, I changed this to (1) assert_uint32 and (2) it does a return 1 not an exit 1.

Yes, I suggested both *independently* and gave possible reasons for doing either. You forgo the reasoning, do both and tell me that I should be happy now in a "I did what you said even though I think it's stupid. Aren't you happy now?" kind of way.

I don't know about you but I'm done talking about bike-sheds and red herrings.
Please do the rest of this without me.